### DISCUSSION TITLE
Can we build and publish an index on macOS machine?

### DISCUSSION THREAD

**Discussion by tmokmss** [awslabs/soci-snapshotter#733] [Q&A]
Hi, as I read [the getting started doc](https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#install-soci-snapshotter), it was not clear if we can use macOS to build an index for a Linux container image.

Does soci-snapshotter only support Linux? Or can we also use macOS (or even Windows) to build an index for a Linux image?

**Comment by sparr** [comment]
We do not currently test on MacOS, but at first glance it seems all of the compile and runtime requirements can be met. The same goes for Windows, particularly using WSL. Please let us know if you attempt and are successful or not.

**Reply by tmokmss** [reply]
Thanks! I'll try it later. Also, official support for those OS will be much appreciated, e.g. to build and publish indices from tools like AWS CDK ([ref](https://github.com/aws/aws-cdk/issues/26413)).

**Reply by Kern--** [reply]
One issue you might run in to is getting the SOCI CLI access to the container image. Tools like Docker desktop manage containers inside a VM with an agent that proxies CLI requests on macOS to the docker engine inside the VM. If that's what CDK is using, then it might not help to have native macOS SOCI binaries because they wouldn't be able to access the image inside the VM.

A [Docker extension](https://www.docker.com/products/extensions/) might be able to help here, but at the moment SOCI only understands how to find container images with containerd. That can be done with Docker desktop, but it's [currently an experimental feature](https://www.docker.com/blog/extending-docker-integration-with-containerd/).

I do think that supporting macOS in some way is a good idea, but I'm not sure that native binaries will solve the real problems macOS users will have.

**Reply by tmokmss** [reply]
As the the ease of setup is crucial for better CDK user experience, that sounds a difficult approach to adopt If it requires special setup steps other than ordinary Docker desktop (or maybe Finch) installation.

I'll consider [the Lambda function approach](https://github.com/aws-ia/cfn-ecr-aws-soci-index-builder/blob/main/functions/source/soci-index-generator-lambda/handler.go) for this specific use case too. At the same time I guess a simpler way to build indices on mac or windows will still be desired to spread your SOCI solution. Thanks!
